packages/base-driver/lib/jsonwp-proxy/proxy.js

```diff
@@ -140,67 +123,61 @@ class JWProxy {
     return this._downstreamProtocol;
   }
 
-  getUrlForProxy(url) {
-    if (url === '') {
-      url = '/';
-    }
-    const proxyBase = `${this.scheme}://${this.server}:${this.port}${this.base}`;
-    const endpointRe = '(/(session|status|appium))';
-    let remainingUrl = '';
-    if (/^http/.test(url)) {
-      const first = new RegExp(`(https?://.+)${endpointRe}`).exec(url);
-      if (!first) {
-        throw new Error('Got a complete url but could not extract JWP endpoint');
-      }
-      remainingUrl = url.replace(first[1], '');
-    } else if (new RegExp('^/').test(url)) {
-      remainingUrl = url;
-    } else {
-      throw new Error(`Did not know what to do with url '${url}'`);
-    }
-
-    const stripPrefixRe = new RegExp('^.*?(/(session|status|appium).*)$');
-    if (stripPrefixRe.test(remainingUrl)) {
-      remainingUrl = /** @type {RegExpExecArray} */ (stripPrefixRe.exec(remainingUrl))[1];
+  /**
+   *
+   * @param {string} url
+   * @param {string} [method]
+   * @returns {string}
+   */
+  getUrlForProxy(url, method) {
+    const parsedUrl = nodeUrl.parse(url || '/');
+    if (
+      !parsedUrl.href || !parsedUrl.pathname
+      || (parsedUrl.protocol && !['http:', 'https:'].includes(parsedUrl.protocol))
+    ) {
+      throw new Error(`Did not know how to proxy the url '${url}'`);
     }
-
-    if (!new RegExp(endpointRe).test(remainingUrl)) {
-      remainingUrl = `/session/${this.sessionId}${remainingUrl}`;
+    const pathname = this.reqBasePath && parsedUrl.pathname.startsWith(this.reqBasePath)
+      ? parsedUrl.pathname.replace(this.reqBasePath, '')
+      : parsedUrl.pathname;
+    const match = COMMAND_WITH_SESSION_ID_MATCHER(pathname);
+    const normalizedPathname = _.trimEnd(
+      match && _.isArray(match.params?.command)
+        ? `/${match.params.command.join('/')}`
+        : pathname,
+      '/'
+    );
+    const commandName = normalizedPathname
+      ? routeToCommandName(
+        normalizedPathname,
+        /** @type {import('@appium/types').HTTPMethod | undefined} */ (method)
+      )
+      : '';
+    const requiresSessionId = !commandName || (commandName && isSessionCommand(commandName));
+    const proxyPrefix = `${this.scheme}://${this.server}:${this.port}${this.base}`;
+    let proxySuffix = normalizedPathname ? `/${_.trimStart(normalizedPathname, '/')}` : '';
+    if (parsedUrl.search) {
+      proxySuffix += parsedUrl.search;
     }
```

Comment:
do you know if we have this test case? I didn't come up with this query param case, so would be nice to have this pattern as well

Reply:
added a test